Enable float requantization for avgpool/gavgpool ops.

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#37037 Enable float requantization for avgpool/gavgpool ops.**

Summary:
For avgpool and gavgpool change requantization scheme:
Similar to conv and linear now convert the accumulated int32
values to float, apply requantization scale which includes the averaging
multiplier.
Conver the resulting float value to int32.
Add output_zero_point.

Benchmark numbers compared to baseline:
% speedup on pixel XL.
------------------------------
|                 | aarch32 | aarch64|
|avgpool   | .4            | 13.6       |
|gavgpool | -2.6%    | 3.5%      |
-------------------------------

Test Plan:
Tested via q8avgpool-test, q8gavgpool-test, average-pooling-test and
global-average-pooling-test in PT QNNPACK.
Also via integated test_quantized.py.
python test/quantization/test_quantized.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21168981](https://our.internmc.facebook.com/intern/diff/D21168981)

kimishpatel committed:
Enable float requantization for avgpool/gavgpool ops.

Summary:
For avgpool and gavgpool change requantization scheme:
Similar to conv and linear now convert the accumulated int32
values to float, apply requantization scale which includes the averaging
multiplier.
Conver the resulting float value to int32.
Add output_zero_point.

Test Plan:
Tested via q8avgpool-test, q8gavgpool-test, average-pooling-test and
global-average-pooling-test in PT QNNPACK.
Also via integated test_quantized.py.
python test/quantization/test_quantized.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented:
LG, does this not require any changes in QNNPACK unit tests?

kimishpatel committed:
Update on "Enable float requantization for avgpool/gavgpool ops."


Summary:
For avgpool and gavgpool change requantization scheme:
Similar to conv and linear now convert the accumulated int32
values to float, apply requantization scale which includes the averaging
multiplier.
Conver the resulting float value to int32.
Add output_zero_point.

Benchmark numbers compared to baseline:
% speedup on pixel XL.
------------------------------
|                 | aarch32 | aarch64|
|avgpool   | .4            | 13.6       |
|gavgpool | -2.6%    | 3.5%      |
-------------------------------

Test Plan:
Tested via q8avgpool-test, q8gavgpool-test, average-pooling-test and
global-average-pooling-test in PT QNNPACK.
Also via integated test_quantized.py.
python test/quantization/test_quantized.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21168981](https://our.internmc.facebook.com/intern/diff/D21168981)

[ghstack-poisoned]

kimishpatel commented:
@supriyar can you take a look at this PR?

supriyar commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/qnnpack/requantization.h:
```diff
@@ -278,31 +260,13 @@ pytorch_qnnp_compute_scalar_avgpool_quantization_params(
-      (int32_t)(uint32_t)output_min - (int32_t)(uint32_t)output_zero_point;
-  params.scalar.output_max_less_zero_point =
-      (int32_t)(uint32_t)output_max - (int32_t)(uint32_t)output_zero_point;
+  params.scalar.scale = scale;
```
Not sure, but do we have assertions anywhere in the code that require the scale to be <1 ? Or did we only have them for conv/linear?

kimishpatel replied:
Yes only for conv/linear. Add and avgpool had different constraints. Not < 1. But for conv/linear I missed them in the integrated code. So I need to fix it.

supriyar reviewed (approved):
Overall logic looks fine, small comment to check scale assertions. 
Haven't reviewed the intrinsics code, so hopefully it is thoroughly tested.

kimishpatel committed:
Update on "Enable float requantization for avgpool/gavgpool ops."


Summary:
For avgpool and gavgpool change requantization scheme:
Similar to conv and linear now convert the accumulated int32
values to float, apply requantization scale which includes the averaging
multiplier.
Conver the resulting float value to int32.
Add output_zero_point.

Benchmark numbers compared to baseline:
% speedup on pixel XL.
------------------------------
|                 | aarch32 | aarch64|
|avgpool   | .4            | 13.6       |
|gavgpool | -2.6%    | 3.5%      |
-------------------------------

Test Plan:
Tested via q8avgpool-test, q8gavgpool-test, average-pooling-test and
global-average-pooling-test in PT QNNPACK.
Also via integated test_quantized.py.
python test/quantization/test_quantized.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21168981](https://our.internmc.facebook.com/intern/diff/D21168981)

[ghstack-poisoned]
